sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeDirectoryAsyncClient.java

```diff
@@ -0,0 +1,464 @@
+import com.azure.storage.blob.specialized.BlockBlobAsyncClient;
+import com.azure.storage.blob.specialized.SpecializedBlobClientBuilder;
+import com.azure.storage.common.implementation.StorageImplUtils;
+import com.azure.storage.file.datalake.implementation.models.PathResourceType;
+import com.azure.storage.file.datalake.models.DataLakeRequestConditions;
+import com.azure.storage.file.datalake.models.PathHttpHeaders;
+import com.azure.storage.file.datalake.models.PathInfo;
+import com.azure.storage.file.datalake.models.PathItem;
+import reactor.core.publisher.Mono;
+
+import java.util.Map;
+
+import static com.azure.core.implementation.util.FluxUtil.monoError;
+import static com.azure.core.implementation.util.FluxUtil.withContext;
+
+/**
+ * This class provides a client that contains directory operations for Azure Storage Data Lake. Operations provided by
+ * this client include creating a directory, deleting a directory, renaming a directory, setting metadata and
+ * http headers, setting and retrieving access control, getting properties and creating and deleting files and
+ * subdirectories.
+ *
+ * <p>
+ * This client is instantiated through {@link PathClientBuilder} or retrieved via
+ * {@link FileSystemAsyncClient#getDirectoryAsyncClient(String) getDirectoryAsyncClient}.
+ *
+ * <p>
+ * Please refer to the <a href=https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction?toc=%2fazure%2fstorage%2fblobs%2ftoc.json>Azure
+ * Docs</a> for more information.
+ */
+public final class DataLakeDirectoryAsyncClient extends PathAsyncClient {
+
+    private final ClientLogger logger = new ClientLogger(DataLakeDirectoryAsyncClient.class);
+
+    /**
+     * Package-private constructor for use by {@link PathClientBuilder}.
+     *
+     * @param pipeline The pipeline used to send and receive service requests.
+     * @param url The endpoint where to send service requests.
+     * @param serviceVersion The version of the service to receive requests.
+     * @param accountName The storage account name.
+     * @param fileSystemName The file system name.
+     * @param directoryName The directory name.
+     * @param blockBlobAsyncClient The underlying {@link BlobContainerAsyncClient}
+     */
+    DataLakeDirectoryAsyncClient(HttpPipeline pipeline, String url, DataLakeServiceVersion serviceVersion,
+        String accountName, String fileSystemName, String directoryName, BlockBlobAsyncClient blockBlobAsyncClient) {
+        super(pipeline, url, serviceVersion, accountName, fileSystemName, directoryName, blockBlobAsyncClient);
+    }
+
+    DataLakeDirectoryAsyncClient(PathAsyncClient pathAsyncClient) {
+        super(pathAsyncClient.getHttpPipeline(), pathAsyncClient.getPathUrl(), pathAsyncClient.getServiceVersion(),
+            pathAsyncClient.getAccountName(), pathAsyncClient.getFileSystemName(), pathAsyncClient.getObjectPath(),
+            pathAsyncClient.getBlockBlobAsyncClient());
+    }
```

Comment:
In the File client, we had method for getFileName(), etc. Do we want the same methods on directory?

Reply:
whoops missed that - its on DirectoryClient too